docs/getting-started/rulesets.md

```diff
@@ -199,3 +198,141 @@ rules:
 The example above will run the recommended rules from the `spectral:oas2` ruleset, but report `operation-2xx-response` as a warning rather than as an error (as is the default behavior in the `spectral:oas2` ruleset).
 
 Available severity levels are `error`, `warn`, `info`, `hint`, and `off`.
+
+
+### Adding a custom function
+
+If the built-in functions are not enough, Spectral gives you a possibility of providing your own ones.
+Previously, this used to be possible only in case of programmatic usage of Spectral, but there is no such limitation anymore,
+and functions are an integral part of Spectral rulesets.
+
+A custom function might be any JS function compliant with [IFunction](https://github.com/stoplightio/spectral/blob/90a0864863fa232bf367a26dace61fd9f93198db/src/types/function.ts#L3#L8) type.
+
+```ts
+export type IFunction<O = any> = (
+  targetValue: any,
+  options: O,
+  paths: IFunctionPaths,
+  otherValues: IFunctionValues,
+) => void | IFunctionResult[];
+```
+
+#### Introduction
+
+It takes exactly the same arguments as built-in functions do, so you are more than welcome to take a look at the existing implementation.
+
+Before you write your first custom function, you must be wary of certain limitations that will apply.
+
+First of all, the custom function must be synchronous.
+
+Apart from the above, Spectral is meant to support a variety of environments, so ideally your function should have an equal or at least very similar functionality in Node.js, browsers etc.
+That said, that's one of the core restriction, namely: do not rely on globals or functions specific to one particular environment.
+If you need to access environment specific API, make sure you provide an alternative for other environments.
+A good example of such situation is `fetch` - a function available natively in a browser context, but missing in Node.js.
```

Comment:
We're already loading a number of really useful utilities for core spectral stuff - should we just expose some of these libraries to custom functions as a libs argument or something, rather than requiring them to import and bundle them with each function? Stuff like ajv, fetch, lodash, and whatever else. This would at least cover the obvious helpful utils - if they need more exotic stuff they'd still have to import and bundle it etc.

Would also greatly reduce the bundled ruleset file, since those common util libraries would no longer need to be bundled in.

Reply:
> We're already loading a number of really useful utilities for core spectral stuff - should we just expose some of these libraries to custom functions as a libs argument or something, rather than requiring them to import and bundle them with each function? Stuff like ajv, fetch, lodash, and whatever else. This would at least cover the obvious helpful utils - if they need more exotic stuff they'd still have to import and bundle it etc.

I'm very skeptical about that. 
A couple of issues I see with this approach:
- code might break anytime we bump the dependency. While, we update the code on our own, we cannot assume the end-developer will do the same every time we happen to do it.
- we will lose flexibility and won't be able to get rid of lodash, ajv, fetch and so on without releasing a breaking release - Studio users will face lots of troubles if their custom functions use any of those deps.

I'm happy to let users create a library bundle that could be injected to each function, but I'm against injecting our own dependencies for others to use.

Reply:
Yeah let's keep our internal utilities internal. 

I am thinking we should create a `resolve` command, but we can talk about that elsewhere.